Library/Homebrew/cmd/update.rb

```diff
@@ -122,25 +122,41 @@ class Updater
 
   def initialize(repository)
     @repository = repository
+    @stashed = false
   end
 
   def pull!
-    safe_system "git", "checkout", "-q", "master"
+    quiet = []
+    quiet << "--quiet" unless ARGV.verbose?
+
+    unless system "git", "diff", "--quiet"
+      puts "Stashing your changes:"
+      system "git", "status", "--short", "--untracked-files"
+      safe_system "git", "stash", "save", "--include-untracked", *quiet
+      @stashed = true
+    end
+
+    safe_system "git", "checkout", "master", *quiet
 
     @initial_revision = read_current_revision
 
     # ensure we don't munge line endings on checkout
     safe_system "git", "config", "core.autocrlf", "false"
 
-    args = ["pull"]
-    args << "--rebase" if ARGV.include? "--rebase"
-    args << "-q" unless ARGV.verbose?
-    args << "origin"
+    args = ["pull", "--rebase", "origin"]
     # the refspec ensures that 'origin/master' gets updated
     args << "refs/heads/master:refs/remotes/origin/master"
+    args += quiet
 
     reset_on_interrupt { safe_system "git", *args }
 
+    if @stashed
+      puts "Restored your changes:"
+      safe_system "git", "stash", "pop", *quiet
```

Comment:
That will make the stash more likely to succeed but the checkout more likely to fail.